### PR TITLE
Add fp16 optimizer state

### PR DESCRIPTION
**Pull request by andersonic**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Add option for fp16 optimizer state with Adam. Note that it likely won't converge on serious models

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu [hunk 1]
```diff
@@ -91,14 +91,16 @@ struct AdamFunctor
 
                 if(j < n && j < chunk_size) {
                     float scaled_grad = incoming_g[ii]/grad_scale;
-                    m[j] = b1*incoming_m[ii] + (1-b1)*scaled_grad;
-                    v[j] = b2*incoming_v[ii] + (1-b2)*scaled_grad*scaled_grad;
+                    float momentum = incoming_m[ii] * b1 + (1-b1)*scaled_grad;
```
[the remark this answers is not in the document]
Comment (andersonic) [reply fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu]: Done!

File: fairscale/optim/adam.py [hunk 2]
```diff
@@ -57,10 +63,19 @@ def __init__(
             max_grad_norm: Optional[float] = 0.0,
             amsgrad: Optional[bool] = False,
             mixed_precision: Optional[bool] = False,
+            optim_fp16: Optional[bool] = False,
```
[the remark this answers is not in the document]
Comment (andersonic) [reply fairscale/optim/adam.py]: Done!

File: fairscale/optim/adam.py [hunk 3]
```diff
@@ -62,20 +67,24 @@ def __init__(
             weight_decay: Optional[float] = 0.0,
             max_grad_norm: Optional[float] = 0.0,
             amsgrad: Optional[bool] = False,
-            mixed_precision: Optional[bool] = False,
-            optim_fp16: Optional[bool] = False,
+            precision: Optional[Union[None, Precision]] = None,
```
Comment (shruti-bh) [inline fairscale/optim/adam.py:70]: Isn't `Optionsal[Precision]` enough? Why do we need `Union[None,Precision]`?

Comment (andersonic) [reply fairscale/optim/adam.py]: Yes, you're totally right! I thought that I needed to allow None since the default is None, but I was mistaken.

**Pull request merged**